### PR TITLE
Allow poly-kinded GADT vars to infer constraints

### LINKED ISSUE #19904: Match type forgets kindness of variable extending wildcarded trait
## Compiler version

3.4.0, 3.4.1-RC1

## Minimized code

```scala
sealed trait Wrap[A <: AnyKind]

trait Foo[T <: AnyKind, X[_ <: T]]

trait Bar[T <: AnyKind, X[_ <: T]] {

  def foo: Foo[
    Wrap[? <: T],
    [x <: Wrap[? <: T]] =>> x match { case Wrap[a] => X[a] }
  ]

}

object Qux extends Bar[Any, [_] =>> Unit] {

  override def foo: Foo[
    Wrap[? <: Any],
    [x <: Wrap[? <: Any]] =>> x match { case Wrap[a] => Unit }
  ] = ???

}
```

## Output

```scala
error overriding method foo in trait Bar of type =>
  Foo[Wrap[?],
    [x <: Wrap[?]] =>>
      x match {
        case Wrap[a] => Unit
      }
  ];
  method foo of type =>
  Foo[Wrap[?],
    [x <: Wrap[?]] =>>
      x match {
        case Wrap[a] => Unit
      }
  ] has incompatible type

Explanation
===========
I tried to show that
  =>
  Foo[Wrap[?],
    [x <: Wrap[?]] =>>
      x match {
        case Wrap[a] => Unit
      }
  ]
conforms to
  =>
  Foo[Wrap[?],
    [x <: Wrap[?]] =>>
      x match {
        case Wrap[a] => Unit
      }
  ]
but none of the attempts shown below succeeded:

  ==> =>   Foo[Wrap[?],     [x <: Wrap[?]] =>>       x match {         case Wrap[a] => Unit       }   ]  <:  =>   Foo[Wrap[?],     [x <: Wrap[?]] =>>       x match {         case Wrap[a] => Unit       }   ]
    ==> Foo[Wrap[?],   [x <: Wrap[?]] =>>     x match {       case Wrap[a] => Unit     } ]  <:  Foo[Wrap[?],   [x <: Wrap[?]] =>>     x match {       case Wrap[a] => Unit     } ]
      ==> [x <: Wrap[?]] =>>   x match {     case Wrap[a] => Unit   }  <:  [x <: Wrap[?]] =>>   x match {     case Wrap[a] => Unit   }
        ==> x match {   case Wrap[a] => Unit }  <:  x match {   case Wrap[a] => Unit }
          ==> Any  <:  x match {   case Wrap[a] => Unit }  = false
          ==> [a] =>> scala.runtime.MatchCase[Wrap[a], Unit]  <:  [a <: AnyKind] =>> scala.runtime.MatchCase[Wrap[a], Unit]
            ==> type bounds [ <: AnyKind]  <:  type bounds []
              ==> AnyKind  <:  Any  = false

The tests were made under the empty constraint
```

## Expectation

This should compile. Notably, the 3rd and 4th lines in the trace show equivalent types (`==> T <: T`).

Replacing the the bound `x <: Wrap[? <: T]` with `x <: (Wrap[? <: T] & Any)` causes the code to compile, even though `Wrap[? <: T]` is always a subtype of `Any` (and so `Wrap[? <: T] & Any` should be equivalent to `Wrap[? <: T]`).

### PR DESCRIPTION
**Pull request by dwijnand**
Allowing the types to flow through in thirdTryNamed requires a tweak to
baseType.  Given `class C[T <: Int]` (from tests/pos-macros/i10864a) the
basetype of a TypeRef of just "C", with the Any class symbol, shouldn't
return Any because "C" is not a proper/value type.  That is, in
baseType, we need to short-circuit TypeProxy.superType (i.e.
underlying), which is returns resultType for a HKLambda.

Fixes #19904

### CODE REVIEW

**Comment by dwijnand** [comment]
With this change, when typing `y match { case Bx[b] => Unit }]`, when we constrain `Bx[b]` against `y`, we correctly infer the `b` goes from `<: AnyKind` to `<: Any`.  With that, when we get to `checkOverride` in RefChecks, we end up with `Any <:< Any` for the lambda param infos.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -580,12 +580,9 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
-            if (tp1.isLambdaSub) return false
-              // Note: We would like to replace this by `if (tp1.hasHigherKind)`
-              // but right now we cannot since some parts of the standard library rely on the
-              // idiom that e.g. `List <: Any`. We have to bootstrap without scalac first.
-            if cls2 eq AnyClass then return true
-            if cls2 == defn.SingletonClass && tp1.isStable then return true
+            if tp1.hasSimpleKind then
+              if cls2 eq AnyClass then return true
+              if cls2 == defn.SingletonClass && tp1.isStable then return true
             return tryBaseType(cls2)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:586]: Was this change actually necessary? Isn't `tryBaseType` always going to fail anyway when `tp1.hasSimpleKind` is `false`? (given that you added the `case HKTypeLambda => false` in `baseType`, which makes sense)

**Comment by sjrd** [comment]
> That is, in baseType, we need to short-circuit TypeProxy.superType (i.e. underlying), which is returns resultType for a HKLambda.

I have always found dubious that `HKTypeLambda.superType` returns `Any`. Would we solve this issue if we made it return `AnyKind` instead?

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:586]: Yes, because then it cascades to more "tries", including allowing a gadt bounded by AnyKind inferring a gadt bound.

**Comment by dwijnand** [comment]
> > That is, in baseType, we need to short-circuit TypeProxy.superType (i.e. underlying), which is returns resultType for a HKLambda.
> 
> I have always found dubious that `HKTypeLambda.superType` returns `Any`. Would we solve this issue if we made it return `AnyKind` instead?

Not sure.  And I wouldn't be surprised it's a battle to make that change.  But we can try.

**Comment by dwijnand** [comment]
> > > That is, in baseType, we need to short-circuit TypeProxy.superType (i.e. underlying), which is returns resultType for a HKLambda.
> > 
> > 
> > I have always found dubious that `HKTypeLambda.superType` returns `Any`. Would we solve this issue if we made it return `AnyKind` instead?
> 
> Not sure. And I wouldn't be surprised it's a battle to make that change. But we can try.

Tried and it was as battle, so I gave up.
